src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2087,7 +2086,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       if (!clazz.isTrait && clazz.isNonBottomSubClass(ConstantAnnotationClass)) {
         val (pConstrOpt, auxConstrs) = body3.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor).splitAt(1)
         for (p <- pConstrOpt) p.symbol.paramss match {
-          case List(ps) =>
+          case List(_) =>
```

Comment:
This is just a weird way of saying `if (p.symbol.paramss.length != 1) CANSA(p, clazz)`. The snippet also suffers from what Diego always points out, creating lists which are discarded, instead of just iterating. There is always a primary ctor.
```
val ctors = body3.iterator.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor)
val p = ctors.next()
if (p.symbol.paramss.length != 1) ConstantAnnotationNeedsSingleArgumentList(p, clazz)
for (c <- ctors) AuxConstrInConstantAnnotation(c, clazz) 
```

Reply:
Yeah I thought about refactoring this but I couldn't 100% convince myself that there is always a primary constructor.